src/lib/y2firstboot/clients/wsl_product_selection.rb

```diff
@@ -0,0 +1,131 @@
+require "registration/yaml_products_reader"
+require "registration/storage"
+
+module Y2Firstboot
+  module Clients
+    # Client for selecting the product to use with WSL (jsc#PED-1380)
+    #
+    # It also allows to indicate whether to install WSL GUI pattern (jsc#PM-3439).
+    class WSLProductSelection < Yast::Client
+      # Runs the client
+      #
+      # @return [Symbol]
+      def run
+        return :next if products.none?
+
+        dialog = Dialogs::WSLProductSelection.new(products,
+          default_product: product,
+          wsl_gui_pattern: wsl_gui_pattern?)
+
+        result = dialog.run
+
+        save(product: dialog.product, wsl_gui_pattern: dialog.wsl_gui_pattern) if result == :next
+
+        result
+      end
+
+    private
+
+      # Saves changes
+      #
+      # @param product [String] Name of the selected product
+      # @param wsl_gui_pattern [Boolean] Whether to install WSL GUI pattern
+      def save(product:, wsl_gui_pattern:)
+        self.product = product
+        self.wsl_gui_pattern = wsl_gui_pattern
+        update_registration
+      end
+
+      # Name of the product to use
+      #
+      # @see {ẂSLConfig}
+      #
+      # @return [String]
+      def product
+        WSLConfig.instance.product || default_product
+      end
+
+      # Sets the product to use
+      #
+      # @see {ẂSLConfig}
+      #
+      # @param value [String] Name of the product
+      def product=(value)
+        WSLConfig.instance.product = value
+      end
+
+      # Whether the WSL GUI pattern should be installed
+      #
+      # @see {ẂSLConfig}
+      #
+      # @return [Boolean]
+      def wsl_gui_pattern?
+        WSLConfig.instance.patterns.include?("wsl_gui")
+      end
+
+      # Sets whether to install the WSL GUI pattern
+      #
+      # @param value [Boolean]
+      def wsl_gui_pattern=(value)
+        if value
+          WSLConfig.instance.patterns.push("wsl_gui").uniq!
+        else
+          WSLConfig.instance.patterns.delete("wsl_gui")
+        end
+      end
+
+      # Updates values stored in registration
+      #
+      # Those values indicates to registration what product was selected and whether the product
+      # has to be registered.
+      #
+      # @see {Registration::Storage::InstallationOptions}
+      def update_registration
+        force_registration = WSLConfig.instance.product_switched? || wsl_gui_pattern?
+        yaml_product = products.find { |p| p["name"] == WSLConfig.instance.product }
+
+        Registration::Storage::InstallationOptions.instance.force_registration = force_registration
+        Registration::Storage::InstallationOptions.instance.yaml_product = yaml_product
+      end
+
+      # Name of the default product to use from YAML file
+      #
+      # @return [String]
+      def default_product
+        return nil if products.none?
+
+        default = products.find { |p| p["default"] } || products.first
+        default["name"]
```

Comment:
I'd save the full product item here, in theory there might be the same product in several different versions. That's unlikely for SLES or SLED, but IIRC we already allowed registering two different versions of SUSE Manager...

Then adapt the selection dialog for this.